engine/access/state_stream/filter.go

```diff
@@ -0,0 +1,89 @@
+package state_stream
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// EventFilter represents a filter applied to events for a given subscription
+type EventFilter struct {
+	hasFilters bool
+	EventTypes map[flow.EventType]struct{}
+	Addresses  map[string]struct{}
+	Contracts  map[string]struct{}
+	EventNames map[string]struct{}
+}
+
+func NewEventFilter(
+	eventTypes []string,
+	addresses []string,
+	contracts []string,
+	eventNames []string,
+) EventFilter {
+	f := EventFilter{
+		EventTypes: make(map[flow.EventType]struct{}, len(eventTypes)),
+		Addresses:  make(map[string]struct{}, len(addresses)),
+		Contracts:  make(map[string]struct{}, len(contracts)),
+		EventNames: make(map[string]struct{}, len(eventNames)),
+	}
+	for _, eventType := range eventTypes {
+		f.EventTypes[flow.EventType(eventType)] = struct{}{}
+	}
+	for _, address := range addresses {
+		// convert to flow.Address to ensure it's in the correct format, but use the string value
+		// for matching to avoid an address conversion for every event
+		f.Addresses[flow.HexToAddress(address).String()] = struct{}{}
+	}
+	for _, contract := range contracts {
+		f.Contracts[contract] = struct{}{}
```

Comment:
this should have validation to ensure contracts are `A.[Address].[ContractName]` or `flow`.